Adding support for protocol handler

marcellintacite opened the pull request:
Here's a draft for this issue : #1247 
This merge request adds support for handling xmpp: protocol links in Converse.js, allowing users to click links like xmpp:user@example.com to automatically open a chat with the specified JID if logged in. This enhances user experience by enabling seamless integration with external XMPP links (e.g., from emails, websites, or other apps).

Before submitting your request, please make sure the following conditions are met:

- [x] Add a changelog entry for your change in `CHANGES.md`
- [ ] When adding a configuration variable, please make sure to
      document it in `docs/source/configuration.rst`
- [ ] Please add a test for your change. Tests can be run in the commandline
      with `make check` or you can run them in the browser by running `make serve`
      and then opening `http://localhost:8000/tests.html`.

marcellintacite commented on src/index.js:
```diff
@@ -42,6 +42,7 @@ import "./plugins/rosterview/index.js";
 import "./plugins/singleton/index.js";
 import "./plugins/dragresize/index.js";     // Allows chat boxes to be resized by dragging them
 import "./plugins/fullscreen/index.js";
+import "./plugins/protocol-handler/index.js"; // Handle xmpp: links
```
I have updated the function

marcellintacite commented on src/plugins/protocol-handler/index.js:
```diff
@@ -0,0 +1,58 @@
+    initialize () {
+        let jidFromUrl = null;  // Store the JID from the URL for later use
+        // The first step is to register the protocol handler on app initialization
+        if ('registerProtocolHandler' in navigator) {
```
hello @jcbrand , i have updated codes, you can take a look

jcbrand commented on src/headless/plugins/chat/utils.js:
```diff
@@ -22,7 +22,16 @@ export function routeToChat (event) {
+
+    // Remove xmpp: prefix if present
+    if (jid.startsWith('xmpp:')) {
+        jid = jid.slice(5);
```
Thanks @marcellintacite, your PR looks better now.

However, there are still various use-cases from XEP-0147 which aren't covered.

For example the message action:
https://xmpp.org/extensions/xep-0147.html#actions-message

And roster management actions:
https://xmpp.org/extensions/xep-0147.html#actions-roster-add

In the very least, you need to remove the query string (the part after `?`) from the xmpp URI, like you removed the `xmpp:` prefix, but ideally you could also implement the actions from XEP-0147.

jcbrand commented on src/plugins/chatview/index.js:
```diff
@@ -65,6 +65,14 @@ converse.plugins.add('converse-chatview', {
 
+        if ('registerProtocolHandler' in navigator) {
+        try {
+            const handlerUrl = `${window.location.origin}${window.location.pathname}#converse/chat?jid=%s`;
```
On second thought, given that XEP-0147 handles more use-cases than just opening a chat with a JID, I think we should perhaps create another URL with its own handler.

Something like ``${window.location.origin}${window.location.pathname}#converse/action?uri=%s``

And then instead of `routeToChat`, the handler could be called `routeToQueryAction` (because XEP-0147 uses the terminology "query actions").

marcellintacite replied:
I have create the function under utils file of the chatview plugin

jcbrand replied:
It would be great if you could also write tests for the work you've done, so that we have more confidence that it's correct and so that future regressions are caught.

To do so, you can create a new file here: https://github.com/conversejs/converse.js/tree/master/src/plugins/chatview/tests

You can call it `route-actions.js`.

I'm not 100% sure what the best approach will be to write the tests, perhaps just calling `routeToQueryActions` manually and then confirming what happens.

If you get stuck or need help, please ask and I'll see if I can help you.

jcbrand commented on manifest.json:
```diff
@@ -26,5 +26,11 @@
+  "protocol_handlers": [
+    {
+      "protocol": "xmpp",
+      "url": "/dev?jid=%s"
```
This URL needs to be updated, right?

marcellintacite replied:
i will update

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,121 @@ export function resetElementHeight (ev) {
+
+        const { jid, queryParams } = parseXMPPURI(uri);
+        if (!u.isValidJID(jid)) {
+            return log.warn(`Invalid JID: "${jid}"`);
```
```suggestion
            return log.warn(`routeToQueryAction: Invalid JID: "${jid}"`);
```

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,121 @@ export function resetElementHeight (ev) {
+
+        const action = queryParams.get('action');
+        if (!action) {
+            log.debug(`No action specified, opening chat for "${jid}"`);
```
```suggestion
            log.debug(`routeToQueryAction: No action specified, opening chat for "${jid}"`);
```

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,121 @@ export function resetElementHeight (ev) {
+                break;
+
+            default:
+                log.warn(`Unsupported XEP-0147 action: "${action}"`);
```
```suggestion
                log.warn(`routeToQueryAction: Unsupported XEP-0147 action: "${action}"`);
```

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,121 @@ export function resetElementHeight (ev) {
+ */
+async function handleRosterAction(jid, params) {
+    await api.waitUntil('connected');
+    await api.waitUntil('rosterContactsFetched');
```
I think you only need the 2nd `waitUntil` because by the time `rosterContactsFetched` has resolved, `connected` will also have resolved. Or is there a specific reason you have both?

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,120 @@ export function resetElementHeight (ev) {
+    let uri = null;
+
+    // Case 1: protocol handler (?uri=...)
+    const searchParams = new URLSearchParams(window.location.search);
```
In this repo, we use `snake_case` to name variables (unless the variable refers to a function) and `camelCase` for functions.

I know this is not how most other JavaScript projects name their variables, but this is an old convention dating back to 2013, when I was influenced by the naming conventions in Python.

To avoid confusion and inconsistency, please stick to the convention of using `snake_case` for variables.

marcellintacite commented on src/plugins/chatview/tests/query-actions.js:
```diff
@@ -0,0 +1,206 @@
+
+        // Mock window.location to simulate a protocol handler URI
+        // This simulates: ?uri=xmpp:romeo@montague.lit
+        delete window.location;
```
Hello @jcbrand , do you know a proper way of working with window.location in jasmine? i have been trying but when i run test i am getting : 
`Failed:` Cannot redefine property: location.` 
I am blocked with tests, i'll need your help.

marcellintacite commented:
Hello @jcbrand i have fixed test issue

jcbrand commented on src/plugins/chatview/tests/query-actions.js:
```diff
@@ -0,0 +1,167 @@
+     * Test handling of invalid JIDs
+     * This ensures the function gracefully handles malformed JIDs
+     */
+    fit("handles invalid JID gracefully",
```
@marcellintacite Please rename these functions so that they're all `it`

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,117 @@ export function resetElementHeight (ev) {
+export async function routeToQueryAction(event) {
+    const { u } = _converse.env;
+
+    try {
```
@marcellintacite  Why is it necessary to put everything in a `try/catch` here?

What happens when it's not there? Generally speaking, a try/catch should only be put around a single line or a small section of code that might be reasonably expected to throw an error that needs to be handled.

jcbrand reviewed (changes requested):
Hi @marcellintacite 

I took some time to review this PR. Thanks a lot for the work you've done on this so far.

Please have a look at this comment first:
https://github.com/conversejs/converse.js/pull/3870/changes#r2667473276

Since it's the most important one of my review.

It's important that we stick to the plugin architecture and separation of concerns, which is why I think you need to move the functionality to separate plugins.

The roster-related stuff should go into the `rosterview` plugin and the other actions can go into `chatboxviews` (which encompasses both MUC and 1:1 whereas `chatview` is only 1:1).

Let me know if you have any questions or if anything is unclear.

jcbrand commented:
@marcellintacite Can you please squash and rebase onto the master branch, then I'll review? Thanks!

marcellintacite commented:
Hi @jcbrand , I have squashed and rebased my commits onto the master branch. I will now await your review.

jcbrand commented on src/plugins/rosterview/utils.js:
```diff
@@ -488,3 +488,54 @@ Object.assign(u, {
+        // Check if user is connected before attempting to add contact
+        if (!api.connection.connected()) {
+            api.alert('warning', __('Not connected'), 
+                [__('You must be connected to add contacts to your roster.')]);
```
```suggestion
                [__('Please login first before adding contacts to your roster.')]);
```

jcbrand commented on src/plugins/chatview/utils.js:
```diff
@@ -71,3 +73,71 @@ export function resetElementHeight (ev) {
+ * window.location.hash = '#converse/action?uri=xmpp%3Auser%40example.com';
+ * await routeToQueryAction();
+ */
+export async function routeToQueryAction(event) {
```
I don't see where this function is ever called.

In the tests you call it manually, which is not useful. This function should be automatically called when the URI changes.

I'm also not convinced that this function belongs in the chatview plugin, since it provides more general functionality relied upon by other plugins.

jcbrand commented:
Thanks @marcellintacite, please see my review comments.

jcbrand commented:
@marcellintacite 

CI is failing

```
> tsc -p ./src/log/tsconfig.json && tsc -p ./src/headless/tsconfig.json && tsc

Error: src/plugins/rosterview/index.js(7,10): error TS2300: Duplicate identifier 'clearXMPPProvidersCache'.
Error: src/plugins/rosterview/index.js(7,35): error TS2300: Duplicate identifier 'highlightRosterItem'.
Error: src/plugins/rosterview/index.js(9,10): error TS2300: Duplicate identifier 'clearXMPPProvidersCache'.
Error: src/plugins/rosterview/index.js(9,35): error TS2300: Duplicate identifier 'highlightRosterItem'.
make: *** [Makefile:248: check] Error 2
/usr/bin/bash /home/runner/work/_actions/GabrielBB/xvfb-action/v1/cleanup.sh
No xvfb processes to kill
Error: The process '/usr/bin/xvfb-run' failed with exit code 2

```

marcellintacite commented:
> @marcellintacite
> 
> CI is failing
> 
> ```
> > tsc -p ./src/log/tsconfig.json && tsc -p ./src/headless/tsconfig.json && tsc
> 
> Error: src/plugins/rosterview/index.js(7,10): error TS2300: Duplicate identifier 'clearXMPPProvidersCache'.
> Error: src/plugins/rosterview/index.js(7,35): error TS2300: Duplicate identifier 'highlightRosterItem'.
> Error: src/plugins/rosterview/index.js(9,10): error TS2300: Duplicate identifier 'clearXMPPProvidersCache'.
> Error: src/plugins/rosterview/index.js(9,35): error TS2300: Duplicate identifier 'highlightRosterItem'.
> make: *** [Makefile:248: check] Error 2
> /usr/bin/bash /home/runner/work/_actions/GabrielBB/xvfb-action/v1/cleanup.sh
> No xvfb processes to kill
> Error: The process '/usr/bin/xvfb-run' failed with exit code 2
> ```

Thanks, i am fixing it

marcellintacite commented:
Hi @jcbrand ,

I've just pushed a major update to this PR to bring the implementation into strict compliance with the [XEP-0147 ](https://xmpp.org/extensions/xep-0147.html) specification.

The previous approach to parsing URIs and routing actions was not fully conforming to the standard. Specifically, the old parser was treating the action as a standard key-value query parameter (e.g., ?action=message&body=Text), whereas XEP-0147 explicitly dictates a semicolon-separated format where the action is the leading key (e.g., ?message;body=Text).

Because of this discrepancy, several core actions were failing silently or behaving incorrectly:
- Action Routing Fixed: Rewrote the URI parser and the xmppURIAction event payload to correctly isolate the leading action type from the subsequent query_params
- Draft Message Pre-filling: Because the old routing was broken, ?message links were opening empty chat windows 
- The newly compliant router now correctly extracts the body parameter and injects it as a draft immediately when the chat box is spawned
- Race Conditions Resolved: Added await api.waitUntil('chatBoxesFetched') across all routers to ensure Converse waits for authentication and state hydration before executing URI actions, preventing startup crashes when clicking external links
- Tests Updated: Completely rewrote the mocked URIs in our test suites to strictly enforce the XEP-0147 semicolon syntax, ensuring our automated tests accurately reflect the specification moving forward.

_Note regarding CI Failures: The TypeScript errors currently failing the CI pipeline (e.g., Property 'resolve' does not exist on type 'Promise<void>' and Type 'typeof PersistentStorage' is missing the following properties...) are originating from upstream changes in the @converse/skeletor 3.1.0 upgrade that this branch recently inherited from master. These errors are entirely unrelated to the XEP-0147 implementation in this PR and will require a separate fix on master to resolve the type discrepancies._

Please kindly review the actual implementation

marcellintacite committed:
feat: Add XEP-0147 protocol handler support for XMPP URIs

jcbrand commented:
Thanks for this, @marcellintacite, and I apologise for the timing.

XEP-0147 URI handling landed on master in ae93376ba on 5 August, covering the same six query actions you did in this PR.

What you did in this PR that wasn't yet in master is registering Converse as the browser's handler so that links clicked
outside Converse open inside it. I refactored out that code in a new commit with you as co-author, and it closes #1247.
